### DISCUSSION TITLE
How can I receive uploadFile and send to minio server without write to temp folder?

### DISCUSSION THREAD

**Discussion by jet10000** [fastapi/fastapi#8073] [Questions] [answered]
```
from minio import Minio
minioClient = Minio('127.0.0.1:9000',
                access_key='xxxx',
                secret_key='xxxx',
                secure=False)

@app.post('/upload/')
async def upload_file(file: UploadFile = File(...)):
    ret = minioClient.fput_object('test', file.filename, file)
    return {'ret': ret}
```

site-packages\minio\api.py", line 606, in fput_object
    with open(file_path, 'rb') as file_data:
TypeError: expected str, bytes or os.PathLike object, not UploadFile

https://docs.min.io/docs/python-client-quickstart-guide.html
https://docs.min.io/docs/python-client-api-reference#fput_object
https://docs.min.io/docs/python-client-api-reference#put_object

```
ret = minioClient.put_object('test', file.filename', file.file, 100)
```
or, How to check the UploadFile size?

# Solved, Any suggestions?
```
file_size = os.fstat(file.file.fileno()).st_size
ret = minioClient.put_object('test', file.filename, file.file, file_size)
```
## Updated
```
minioClient.fput_object('test', file.filename, file.file.fileno())
```

**Comment by tiangolo** [comment, marked as answer]
Thanks for reporting back with your solution!

That's correct. The `file.file` contains the actual Python file object, so you can use that directly with Minio.

The only thing missing is, use a normal `def` function for your path operation, not `async def`. This is because Minio uses I/O (network) but doesn't use `async` and `await`.

**Comment by parsa10** [comment]
> ```
> from minio import Minio
> minioClient = Minio('127.0.0.1:9000',
>                 access_key='xxxx',
>                 secret_key='xxxx',
>                 secure=False)
> 
> @app.post('/upload/')
> async def upload_file(file: UploadFile = File(...)):
>     ret = minioClient.fput_object('test', file.filename, file)
>     return {'ret': ret}
> ```
> 
> site-packages\minio\api.py", line 606, in fput_object with open(file_path, 'rb') as file_data: TypeError: expected str, bytes or os.PathLike object, not UploadFile
> 
> https://docs.min.io/docs/python-client-quickstart-guide.html https://docs.min.io/docs/python-client-api-reference#fput_object https://docs.min.io/docs/python-client-api-reference#put_object
> 
> ```
> ret = minioClient.put_object('test', file.filename', file.file, 100)
> ```
> 
> or, How to check the UploadFile size?
> 
> # Solved, Any suggestions?
> ```
> file_size = os.fstat(file.file.fileno()).st_size
> ret = minioClient.put_object('test', file.filename, file.file, file_size)
> ```
> 
> ## Updated
> ```
> minioClient.fput_object('test', file.filename, file.file.fileno())
> ```

It's working , thanks a lot

**Reply by RichieB2B** [reply]
Please note that using `file.file.fileno()` still creates a temporary file on disk because Starlette `UploadFile` will use `SpooledTemporaryFile` and `rollover()` the data to disk via `TemporaryFile` to create a `fileno()`. In a read-only pod this results in:
```
FileNotFoundError: [Errno 2] No usable temporary directory found in ['/tmp', '/var/tmp', '/usr/tmp', '/usr/app']
```
IMHO the best in-memory solution is:
```
@app.post('/upload/')
def upload_file(file: UploadFile = File(...)):
    ret = minioClient.put_object('test', file.filename, file.file, file.size)
    return {'ret': ret}
```
